<div style="max-width:66ch;">

# Exercise - pandas fundamentals

---
These are introductory exercises in Pandas with focus in **syntax, DataFrame, indexing, data selection** and **data cleaning**.

<p class = "alert alert-info" role="alert"><b>Note</b> that there are usually many ways to achieve same results using operations in pandas. Feel free to investigate several ways. </p>

<p class = "alert alert-info" role="alert"><b>Note</b> that in cases when you start to repeat code, try not to. Create functions to reuse code instead. </p>

<p class = "alert alert-info" role="alert"><b>Remember</b> to use <b>descriptive variable, function, index </b> and <b> column names</b> in order to get readable code </p>

<p class = "alert alert-info" role="alert"><b>Remember</b> to format your input questions in a pedagogical way to guide the user

The number of stars (\*), (\*\*), (\*\*\*) denotes the difficulty level of the task

---

</div>

<div style="max-width:66ch;">

## 0. Cities in Sweden - create dataset (*)

Create this DataFrame from scratch: 

|    | Kommun    |   Population |
|---:|:----------|-------------:|
|  0 | Malmö     |       347949 |
|  1 | Stockholm |       975551 |
|  2 | Uppsala   |       233839 |
|  3 | Göteborg  |       583056 |

&nbsp; a) Use your DataFrame to print out all the cities. (*)

&nbsp; b) Select only the row which contains Göteborg. Do this by using the name Göteborg. (*)

&nbsp; c) Sort the cities by population from largest to smallest. (*)

&nbsp; d) Filter out the three largest cities. (*)

&nbsp; e) The whole population in Sweden 2020 is 10379295. Use this number to create a new column in your sorted DataFrame named: Population (%). This column should be filled with percentage of the Swedish population for each city. 

</div>




In [39]:
# a) Use your DataFrame to print out all the cities. (*)import pandas as pd

cities = {
"Kommun": ["Malmö","Stockholm","Uppsala","Göteborg"],
"Population":[347949,975551,233839,583056]
}

df = pd.DataFrame(cities)

print(df)

      Kommun  Population
0      Malmö      347949
1  Stockholm      975551
2    Uppsala      233839
3   Göteborg      583056


In [40]:
# b) Select only the row which contains Göteborg. Do this by using the name Göteborg. (*)

gothenburg = df[df["Kommun"] == "Göteborg"]

print(gothenburg)

     Kommun  Population
3  Göteborg      583056


In [41]:
# c) Sort the cities by population from largest to smallest. (*)

sorted = df.sort_values(by="Population", ascending=False)

print(sorted)

      Kommun  Population
1  Stockholm      975551
3   Göteborg      583056
0      Malmö      347949
2    Uppsala      233839


In [42]:
# d) Filter out the three largest cities. (*)

sorted = df.sort_values(by="Population", ascending=False)

top_three = sorted.head(3)

print(top_three)

      Kommun  Population
1  Stockholm      975551
3   Göteborg      583056
0      Malmö      347949


<div style="max-width:66ch;">

---


## 1. Cities in Sweden - real dataset (*)


Download the file komtopp50_2020.xlsx from the course github repo. The data is taken from SCB. 

&nbsp; a) Read in the tab "Totalt" into a DataFrame and start exploring the data with some simple explorations such as

```python
df.head()
df.info()
df.describe()
```

Feel free to do more explorations. (*)

&nbsp; b) Clean your data so that the head looks like this: (*)


|    |   Rang 2020 |   Rang 2019 | Kommun   |   Folkmängd 2020 |   Folkmängd 2019 |   Förändring |
|---:|------------:|------------:|:---------|-----------------:|-----------------:|-------------:|
|  0 |          83 |          84 | Ale      |            31868 |            31402 |     1.48398  |
|  1 |          64 |          64 | Alingsås |            41602 |            41420 |     0.439401 |
|  2 |         123 |         123 | Alvesta  |            20224 |            20134 |     0.447005 |
|  3 |         255 |         255 | Aneby    |             6821 |             6848 |    -0.394276 |
|  4 |         169 |         167 | Arboga   |            14039 |            14087 |    -0.34074  |

&nbsp; c) Sort the cities by population from largest to smallest. (*)

&nbsp; d) Filter out the five smallest cities. (*)

&nbsp; e) Use the DataFrame to calculate Sweden's population in 2019 and 2020. (*)

&nbsp; f) Plot a bar chart for the five largest cities and the five smallest cities. (*) 


</div>




In [ ]:
# a) Read in the tab "Totalt"

import pandas as pd

# Fick använda en absolute filepath
df = pd.read_excel("C:/Users/lager/OneDrive/Documents/Github/Python 1/Python-Bjorn-Lagerblad-OPA23/data/data_processing/komtopp50_2020.xlsx", sheet_name="Totalt")


print("First few rows:")
print(df.head())

print("\nDataFrame info:")
print(df.info())

print("\nSummary statistics:")
print(df.describe())




In [52]:
# b) Clean your data.

df = df.dropna()

df = df.reset_index(drop=True)

df['Kommun'] = df['Kommun'].str.strip()

print("Cleaned data head:")
print(df.head())


Cleaned data head:
  Rang 2020  Rang 2019    Kommun Folkmängd 2020  Folkmängd 2019 Förändring
0        83       84.0       Ale          31868         31402.0   1.483982
1        64       64.0  Alingsås          41602         41420.0   0.439401
2       123      123.0   Alvesta          20224         20134.0   0.447005
3       255      255.0     Aneby           6821          6848.0  -0.394276
4       169      167.0    Arboga          14039         14087.0   -0.34074


In [54]:
# c) Sort the cities by population from largest to smallest.

df_sorted = df.sort_values(by="Folkmängd 2020", ascending=False)

print("Largest to smallest:")
print(df_sorted)

Largest to smallest:
    Rang 2020  Rang 2019     Kommun Folkmängd 2020  Folkmängd 2019 Förändring
198         1        1.0  Stockholm         975551        974073.0   0.151734
57          2        2.0   Göteborg         583056        579281.0    0.65167
134         3        3.0      Malmö         347949        344166.0   1.099179
241         4        4.0    Uppsala         233839        230767.0   1.331213
124         5        5.0  Linköping         164616        163051.0   0.959822
..        ...        ...        ...            ...             ...        ...
271       286      286.0      Åsele           2805          2794.0   0.393701
5         287      287.0   Arjeplog           2718          2785.0  -2.405745
29        288      288.0    Dorotea           2498          2551.0  -2.077617
194       289      289.0    Sorsele           2442          2489.0  -1.888309
12        290      290.0   Bjurholm           2387          2408.0  -0.872093

[290 rows x 6 columns]


In [60]:
# d) Filter out the five smallest cities. (*)

df["Folkmängd 2020"] = pd.to_numeric(df["Folkmängd 2020"], errors="coerce")

smallest = df.nsmallest(5, "Folkmängd 2020")

print(smallest)

  Rang 2020  Rang 2019    Kommun  Folkmängd 2020  Folkmängd 2019 Förändring
0        83       84.0       Ale             NaN         31402.0   1.483982
1        64       64.0  Alingsås             NaN         41420.0   0.439401
2       123      123.0   Alvesta             NaN         20134.0   0.447005
3       255      255.0     Aneby             NaN          6848.0  -0.394276
4       169      167.0    Arboga             NaN         14087.0   -0.34074


In [2]:
import pandas as pd

file_path = r'C:\Users\lager\OneDrive\Documents\Github\Python 1\Python-Bjorn-Lagerblad-OPA23\data\data_processing'
df = pd.read_excel(file_path)

total_population_2020 = df['Folkmängd 2020'].sum()

print("Total population in 2020:", total_population_2020)


PermissionError: [Errno 13] Permission denied: 'C:\\Users\\lager\\OneDrive\\Documents\\Github\\Python 1\\Python-Bjorn-Lagerblad-OPA23\\data\\data_processing'

<div style="max-width:66ch;">

---

## 2. Cities in Sweden - gender (\*)

We continue with the same Excel-file as in task 2, but now you should also read in the sheets "Kvinnor" and "Män" into two additional DataFrames. In this task, many operations are similar to all three datasets, try creating custom made functions to reuse as much code as possible.

&nbsp; a) Clean your data so that the head looks like this: (\*)

Male:

|     | Rang 2020 | Rang 2019 | Kommun   | Folkmängd 2020 | Folkmängd 2019 | Förändring | Kön |
| --: | --------: | --------: | :------- | -------------: | -------------: | ---------: | :-- |
|   0 |        81 |        83 | Ale      |          16256 |          16054 |    1.25825 | Man |
|   1 |        64 |        64 | Alingsås |          20702 |          20646 |   0.271239 | Man |
|   2 |       122 |       123 | Alvesta  |          10367 |          10339 |   0.270819 | Man |
|   3 |       255 |       255 | Aneby    |           3503 |           3498 |   0.142939 | Man |
|   4 |       169 |       168 | Arboga   |           7094 |           7114 |  -0.281136 | Man |

Female:

|     | Rang 2020 | Rang 2019 | Kommun   | Folkmängd 2020 | Folkmängd 2019 | Förändring | Kön    |
| --: | --------: | --------: | :------- | -------------: | -------------: | ---------: | :----- |
|   0 |        84 |        85 | Ale      |          15612 |          15348 |    1.72009 | Kvinna |
|   1 |        64 |        64 | Alingsås |          20900 |          20774 |   0.606527 | Kvinna |
|   2 |       123 |       123 | Alvesta  |           9857 |           9795 |   0.632976 | Kvinna |
|   3 |       255 |       255 | Aneby    |           3318 |           3350 |  -0.955224 | Kvinna |
|   4 |       165 |       164 | Arboga   |           6945 |           6973 |  -0.401549 | Kvinna |

&nbsp; b) Merge the male and female DataFrames vertically and set index to "Kommun". Note that there now should be 580 rows now. (\*)

|     Kommun | Folkmängd 2020 | Folkmängd 2019 | Förändring |    Kön |
| ---------: | :------------- | -------------: | ---------: | -----: |
|        Ale | 16256          |          16054 |    1.25825 |    Man |
|   Alingsås | 20702          |          20646 |   0.271239 |    Man |
|        ... | ...            |            ... |        ... |    ... |
|  Överkalix | 1559           |           1578 |   -1.20406 | Kvinna |
| Övertorneå | 2027           |           2065 |   -1.84019 | Kvinna |

&nbsp; c) Extract and change column name from the total DataFrame so that the head look like this: (\*)

|     | Kommun   | Total Pop 2020 | Total Pop 2019 | Total förändring |
| --: | :------- | -------------: | -------------: | ---------------: |
|   0 | Ale      |          31868 |          31402 |          1.48398 |
|   1 | Alingsås |          41602 |          41420 |         0.439401 |
|   2 | Alvesta  |          20224 |          20134 |         0.447005 |
|   3 | Aneby    |           6821 |           6848 |        -0.394276 |
|   4 | Arboga   |          14039 |          14087 |         -0.34074 |

&nbsp; d) Merge this data with the data in b) so that the head look like this: (\*)

| Kommun    | Folkmängd 2020 | Folkmängd 2019 | Förändring | Kön    | Total Pop 2020 | Total Pop 2019 | Total förändring |
| :-------- | -------------: | -------------: | ---------: | :----- | -------------: | -------------: | ---------------: |
| Stockholm |         482982 |         482220 |   0.158019 | Man    |         975551 |         974073 |         0.151734 |
| Stockholm |         492569 |         491853 |   0.145572 | Kvinna |         975551 |         974073 |         0.151734 |
| Göteborg  |         292505 |         290308 |   0.756782 | Man    |         583056 |         579281 |          0.65167 |
| Göteborg  |         290551 |         288973 |   0.546072 | Kvinna |         583056 |         579281 |          0.65167 |
| Malmö     |         175411 |         173543 |    1.07639 | Kvinna |         347949 |         344166 |          1.09918 |

&nbsp; e) Create barplots showing the gender populations of Swedens 10 largest and 10 smallest cities. (\*)

&nbsp; f) Create a pie chart showing the total male and female population in Sweden 2020. (\*)

&nbsp; g) Create a barplot showing the cities with the five largest percentual gender difference in 2020. (\*\*)

&nbsp; h) Create a barplot showing the top 5 cities with largest populational growth from 2019 to 2020 (\*\*)

&nbsp; i) Feel free to investigate other questions you are interested in using these datasets. (\*), (\*\*)

</div>


<div style="width: 66ch;">


</div>

<div style="background-color: #FFF; color: #212121; border-radius: 20px; width:25ch; box-shadow: rgba(0, 0, 0, 0.16) 0px 1px 4px; display: flex; justify-content: center; align-items: center;">
<div style="padding: 1em; width: 60%;">
    <h2 style="font-size: 1.2rem;">Kokchun Giang</h2>
    <a href="https://www.linkedin.com/in/kokchungiang/" target="_blank" style="display: flex; align-items: center; gap: .4em; color:#0A66C2;">
        <img src="https://content.linkedin.com/content/dam/me/business/en-us/amp/brand-site/v2/bg/LI-Bug.svg.original.svg" width="20"> 
        LinkedIn profile
    </a>
    <a href="https://github.com/kokchun/Portfolio-Kokchun-Giang" target="_blank" style="display: flex; align-items: center; gap: .4em; margin: 1em 0; color:#0A66C2;">
        <img src="https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png" width="20"> 
        Github portfolio
    </a>
    <span>AIgineer AB</span>
    <div>
</div>